pkg/pillar/types/global.go

```diff
@@ -103,312 +107,417 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
-// If a instance has been removed its read/write vdisks are deleted after
-// one hour.
-var GlobalConfigDefaults = GlobalConfig{
-	ConfigInterval:          60,
-	MetricInterval:          60,
-	ResetIfCloudGoneTime:    7 * 24 * 3600,
-	FallbackIfCloudGoneTime: 300,
-	MintimeUpdateSuccess:    600,
-
-	NetworkGeoRedoTime:        3600, // 1 hour
-	NetworkGeoRetryTime:       600,  // 10 minutes
-	NetworkTestDuration:       30,
-	NetworkTestInterval:       300, // 5 minutes
-	NetworkTestBetterInterval: 0,   // Disabled
-	NetworkFallbackAnyEth:     TS_ENABLED,
-	NetworkTestTimeout:        15,
-
-	NetworkSendTimeout: 120,
-
-	UsbAccess:           true, // Contoller likely to default to false
-	SshAccess:           true, // Contoller likely to default to false
-	SshAuthorizedKeys:   "",
-	StaleConfigTime:     600,  // Use stale config for up to 10 minutes
-	DownloadGCTime:      600,  // 10 minutes
-	VdiskGCTime:         3600, // 1 hour
-	DownloadRetryTime:   600,  // 10 minutes
-	DomainBootRetryTime: 600,  // 10 minutes
-	RktGCGracePeriod:    3600, // 1 hour
-
-	AllowNonFreeAppImages:  TS_ENABLED,
-	AllowNonFreeBaseImages: TS_ENABLED,
-
-	DefaultLogLevel:       "info", // XXX Should we change to warning?
-	DefaultRemoteLogLevel: "info", // XXX Should we change to warning?
-
-	Dom0MinDiskUsagePercent: 20,
-	IgnoreDiskCheckForApps:  false,
+	StringDefault   string
+	BoolDefault     bool
+	TriStateDefault TriState
 }
 
-// Check which values are set and which should come from defaults
-// Zero integers means to use default
-func ApplyGlobalConfig(newgc GlobalConfig) GlobalConfig {
+type ConfigItemSpecMap struct {
+	// GlobalSettings key is the agent name.
+	GlobalSettings map[string]ConfigItemSpec
+	// AgentSettings key is the agent name. The inner map's key is the per-agent setting name.
+	AgentSettings map[string]map[string]ConfigItemSpec
+}
 
-	if newgc.ConfigInterval == 0 {
-		newgc.ConfigInterval = GlobalConfigDefaults.ConfigInterval
-	}
-	if newgc.MetricInterval == 0 {
-		newgc.MetricInterval = GlobalConfigDefaults.MetricInterval
-	}
-	if newgc.ResetIfCloudGoneTime == 0 {
-		newgc.ResetIfCloudGoneTime = GlobalConfigDefaults.ResetIfCloudGoneTime
-	}
-	if newgc.FallbackIfCloudGoneTime == 0 {
-		newgc.FallbackIfCloudGoneTime = GlobalConfigDefaults.FallbackIfCloudGoneTime
+func (configMap *ConfigItemSpecMap) AddIntItem(key string, agentSettingKey string, defaultInt uint32, min uint32, max uint32) {
+	if defaultInt < min || defaultInt > max {
+		log.Fatalf("Adding int item %s failed. Value does not meet given min/max criteria", key)
 	}
-	if newgc.MintimeUpdateSuccess == 0 {
-		newgc.MintimeUpdateSuccess = GlobalConfigDefaults.MintimeUpdateSuccess
+	configItem := ConfigItemSpec{
+		ItemType:   CONFIGITEM_TYPE_INT,
+		Key:        key,
+		IntDefault: defaultInt,
+		IntMin:     min,
+		IntMax:     max,
 	}
-	if newgc.NetworkGeoRedoTime == 0 {
-		newgc.NetworkGeoRedoTime = GlobalConfigDefaults.NetworkGeoRedoTime
+	if agentSettingKey == "" {
```

Comment:
Hmm.. This is confusing.. For agentSetting, you are using key as the agent name? But require agentSettingKey separately?  That is very confusing.. 

I think what I suggested was:
  func (configMap *ConfigItemSpecMap) AddIntItem(key string, globalSetting bool, agentSetting bool, defaultInt uint32, min uint32, max uint32) {
   <Snip.. >

  if globalSetting {
      configMap.GlobalSettings[key] = configItem
  }
  if agentSetting {
      configMap.AgentSettings[key] = configItem
  }
   
}

Reply:
Yes, but @eriknordmark felt that the bools were unnecessary so I removed them